velox/common/memory/MemoryPool.h

```diff
@@ -890,6 +890,142 @@ class MemoryPoolImpl : public MemoryPool {
   // The number of internal memory reservation collisions caused by concurrent
   // memory reservation requests.
   std::atomic<uint64_t> numCollisions_{0};
+
+#define MEMORY_DEBUG
+#ifdef MEMORY_DEBUG
+#define RECORD_ALLOC_DEBUG(...) recordAllocDbg(__VA_ARGS__)
+#define RECORD_FREE_DEBUG(...) recordFreeDbg(__VA_ARGS__)
+#define LEAK_CHECK_DEBUG() leakCheckDbg()
+
+  std::mutex dbgAllocMutex_;
+  // Map from address to {alloc size, call stack}
+  std::unordered_map<uint64_t, std::pair<uint64_t, std::string>>
+      addrToAllocDetail_;
+
+  /// Recording on every allocation is very expensive and normally will result
+  /// in CPU saturation. Modify this method while debugging to limit the number
+  /// of times that the allocations are recorded. 'isAlloc' will be true at
+  /// allocation sites, false at free sites. You want to specify the condition
+  /// so that it is guaranteed that each alloc recording will have a
+  /// corresponding free recording called. In other words, always make sure the
+  /// condition at free sites is less strict than the condition at alloc sites.
+  inline bool allocDbgConditionMet(bool isAlloc) {
```

Comment:
Please remove this and introduce in followup? Thanks!

Reply:
Leaving this here so that it can be directly used (by modifying the condition code). Followup will add feature to allow dynamic modification of the filter condition.